lib/plugin.rb

```diff
@@ -1,10 +1,68 @@
+require 'cocoapods-core'
+require 'keyring_liberator'
+require 'key_master'
+
 module CocoaPodsKeys
 
-  Pod::HooksManager.register('cocoapods-keys', :post_install) do |context, user_options|
-    require 'installer'
-    require 'preinstaller'
+end
+
+module Pod
+  class Installer
+    include Pod::Podfile::DSL
+
+    alias_method :original_install!, :install!
+
+    def install!
+      require 'preinstaller'
+      user_options = config.podfile.plugins["cocoapods-keys"]
+      CocoaPodsKeys::PreInstaller.new(user_options).setup
+
+      # Add our template podspec (needs to be remote, not local). 
+      config.podfile.pod 'Keys', :git => 'https://github.com/ashfurrow/empty-podspec.git', :commit => 'e6588faecb89632f2616251746a313a2ad6336fa'
+
+      original_install!
+    end
+
+    class Analyzer
+      class SandboxAnalyzer
+
+        alias_method :original_pod_state, :pod_state
+
+        def pod_state(pod) 
+          if pod == 'Keys'
+            # return :added if we were, otherwise assume the Keys have :changed since last install, following my mother's "Better Safe than Sorry" principle.
+            return :added if pod_added?(pod)
+            :changed
+          else
+            original_pod_state(pod)
+          end
+        end
+      end
+    end
+  end
+
+  class Specification
+    class << self 
+
+      include Pod::Podfile::DSL
+
+      alias_method :original_from_string, :from_string
+
+      def from_string(spec_contents, path, subspec_name = nil)
+        if path.to_s.include? "Keys.podspec"
+          user_options = Pod::Config.instance.podfile.plugins["cocoapods-keys"]
+
+          keyring = CocoaPodsKeys::KeyringLiberator.get_keyring_named(user_options["project"]) || CocoaPodsKeys::KeyringLiberator.get_keyring(Dir.getwd)
```

Comment:
> Accessing the config from various places in this file (usually for its plugins).

At the very least you can drop the `Pod::` prefix, because you are inside the `module Pod … end` scope. Other than that, it’s gonna be hard to do so, without adding some kind of global, because the code is spread across multiple module/class scopes.